File: src/order-exception-triage/src/main/java/io/ioka/demo/orders/ProviderFixture.java
// SPDX-License-Identifier: Apache-2.0
package io.ioka.demo.orders;

import com.sun.net.httpserver.HttpServer;
import io.ioka.munarium.client.model.Json;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.atomic.*;
import java.util.regex.Pattern;

/** Canned protocol responses derived from supplied evidence; no inference and no oracle access. */
public final class ProviderFixture {
    private ProviderFixture() {}
    public static void run() throws Exception {
        var server = HttpServer.create(new InetSocketAddress(11434), 0);
        var fail = new AtomicBoolean(); var calls = new AtomicInteger();
        server.createContext("/", exchange -> {
            int status = 200; Object response;
            try {
                String path = exchange.getRequestURI().getPath();
                if (path.equals("/api/tags")) response = Map.of("models", List.of(Map.of("name", "order-selected", "model", "order-selected")));
                else if (path.equals("/calls")) response = Map.of("calls", calls.get());
                else if (path.equals("/fail")) { fail.set(true); response = Map.of("ok", true); }
                else if (path.equals("/reset")) { fail.set(false); response = Map.of("ok", true); }
                else if (path.equals("/api/chat")) {
                    calls.incrementAndGet();
                    if (fail.get()) { status = 503; response = Map.of("error", "controlled provider outage"); }
                    else {
                        var body = Json.MAPPER.readTree(exchange.getRequestBody());
                        StringBuilder prompt = new StringBuilder(); for (var message : body.path("messages")) prompt.append(message.path("content").asText()).append('\n');
                        String evidence = prompt.toString().split("EVIDENCE_START")[1].split("EVIDENCE_END")[0];
                        var citation = Pattern.compile("\\[([^\\[\\]\\s]+/[^\\[\\]\\s]+)\\]").matcher(evidence);
                        var route = Pattern.compile("route to ([A-Z_]+)").matcher(evidence);
                        if (!citation.find() || !route.find()) throw new IllegalStateException("Fixture received no procedure evidence.");
                        String explanation = evidence.substring(evidence.indexOf("Procedure revision:")).replaceAll("\\[[^]]+\\]", "").trim();
                        String answer = FilesUtil.json(Map.of("explanation", explanation, "route", route.group(1), "disposition", "review_required", "missing_evidence", evidence.contains("Export documents missing") || evidence.contains("Unknown hold"), "citations", List.of(citation.group(1))));
                        response = Map.of("model", body.path("model").asText(), "done", true, "done_reason", "stop", "message", Map.of("role", "assistant", "content", answer), "prompt_eval_count", 24, "eval_count", 32);
                    }
                } else { status = 404; response = Map.of("error", "unknown route"); }
                byte[] bytes = FilesUtil.json(response).getBytes(StandardCharsets.UTF_8);
                exchange.getResponseHeaders().set("Content-Type", "application/json"); exchange.sendResponseHeaders(status, bytes.length); exchange.getResponseBody().write(bytes);
            } catch (Exception e) { exchange.sendResponseHeaders(500, -1); }
            finally { exchange.close(); }
        });
        server.start();
        System.out.println("Controlled order fixture listening; no model runtime.");
    }
}

File: src/inventory-replenishment/inventory_demo/provider_fixture.py
# SPDX-License-Identifier: Apache-2.0
"""Canned wire fixture; reads only the evidence supplied in the prompt."""

import json
import re
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

STATE = {"mode": "ok", "calls": 0}


class Handler(BaseHTTPRequestHandler):
    def log_message(self, *_args):
        pass

    def respond(self, body, status=200):
        raw = json.dumps(body).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(raw)))
        self.end_headers()
        self.wfile.write(raw)

    def do_GET(self):
        if self.path == "/api/tags":
            self.respond({"models": [{"name": "inventory-fixture", "model": "inventory-fixture"}]})
        elif self.path == "/state":
            self.respond(STATE)
        else:
            self.respond({}, 404)

    def do_POST(self):
        size = int(self.headers.get("Content-Length", "0"))
        if not 0 < size < 1_000_000:
            self.respond({}, 400)
            return
        body = json.loads(self.rfile.read(size))
        if self.path == "/mode":
            STATE["mode"] = body["mode"]
            self.respond({})
            return
        STATE["calls"] += 1
        if STATE["mode"] == "unavailable":
            self.respond({"error": "Controlled provider outage"}, 503)
            return
        prompt = "\n".join(m["content"] for m in body["messages"])
        evidence = re.search(r"evidence: (ev-[A-Za-z0-9-]+)", prompt)
        procedure = re.search(r"\[(procedures/[^\]]+)\]", prompt)
        actions = []
        for line in prompt.splitlines():
            cells = line.split(" | ")
            if len(cells) == 6 and cells[0].startswith("SKU-"):
                actions.append(
                    {
                        "sku": cells[1],
                        "constraint_code": cells[4],
                        "citations": [
                            f"evidence/{evidence.group(1)}#{cells[0]}",
                            procedure.group(1),
                        ],
                    }
                )
        if STATE["mode"] == "irrelevant" and actions:
            actions[0]["citations"][0] = "evidence/ev-unknown#missing"
        answer = {
            "actions": actions,
            "procedure_quote": "Obtain supplier confirmation before requesting replenishment.",
        }
        self.respond(
            {
                "model": body["model"],
                "done": True,
                "done_reason": "stop",
                "message": {"role": "assistant", "content": json.dumps(answer)},
                "prompt_eval_count": 120,
                "eval_count": 60,
            }
        )


if __name__ == "__main__":
    ThreadingHTTPServer(("0.0.0.0", 11434), Handler).serve_forever()
